### PR TITLE
MatNet implementation for FFSP

### PR DESCRIPTION
**Pull request by LTluttmann**
## Description

- FFSPEnv:
	- bug fixes:
		- .view() on run_time messed up the run times (interleaving instead of stacking behavior)
		- reward missed the "-" sign
	- added inline comments
	- implemented render function
	- implemented a mechanism allowing for N POMO starts. Other than the original implementation, this works for arbitrary instances (instead of assuming 3 stages with 4 machines each)
	- implemented a "flatten" option, which flattens out the stage dimension. This allows the use of the (single-stage) MatNet model
	- updated MatNet test

- MatNet
	- implemented the multi-stage MatNet Model for the FFSP as described in the original MatNet paper
	- implemented a single-stage MatNet model for the FFSP. This has the advantage over the original implementation that it is size-agnostic. The multi-stage model can only be applied to instances with a fixed number of stages. The performance of the single-stage model is still under evaluation.
	- added function which, based on the environment, selects the correct MatNet policy

- MatNetCrossMHA:
	- changed some default parameters in accordance to the official paper implementation

- LogitAttention:
	- changed the keyword argument "linear_bias" to "out_bias", since "linear_bias" is already used in AutoregressiveDecoder. Thus, LogitAttention's linear_bias could not be changed without changing that of the AutoregressiveDecoder.

- nn.env_embeddings:
	- added context, init and dynamic emebddings for FFSP
	- replaced the MatNetATSPInitEmbedding with a more generic MatNetInitMebedding (works for ATSP and FFSP)

- POMO: 
	- added "max_reward" to td also during training (for logging purposes)

- RL4COEnvBase:
	- added get_num_starts and select_start_nodes as functions to the environment. By default, the get_num_starts select_start_nodes from the rl4co.utils.ops module are called (w.i.p.), but environments can override these functions with custom logic (example in FFSPEnv).
	- removed select_start_nodes_fn attribute from AutoregressiveDecoder. Note, that it can still be manually set by passing the select_start_nodes_fn argument to the forward function of the decoder (which hands it over to the DecodingStrategy)

- config (main.yaml):
	- moved "metrics" under the model config

## Motivation and Context

- Added working FFSP MatNet implementation
- improvements and fixes for FFSP environment
- let get_num_starts and select_start_nodes functions be implemented by environments as discussed here: https://github.com/ai4co/rl4co/discussions/117
- some minor fixes

- [ ] I have raised an issue to propose this change ([required](https://github.com/ai4co/rl4co/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/envs/common/base.py [hunk 1]
```diff
@@ -173,6 +174,12 @@ def get_action_mask(self, td: TensorDict) -> torch.Tensor:
         """
         raise NotImplementedError
 
+    def get_num_starts(self, td):
+        return get_num_starts(td, self.name)
```
Comment (fedebotu) [inline rl4co/envs/common/base.py:178]: Awesome! 

At the moment this is not yet fully modularized (I guess each env should implement this function)
- In the decoding, i.e. [here](https://github.com/ai4co/rl4co/blob/66622a2e7d616e72696904e1552cf7d61dad817f/rl4co/models/nn/dec_strategies.py#L70) and [here](https://github.com/ai4co/rl4co/blob/66622a2e7d616e72696904e1552cf7d61dad817f/rl4co/models/nn/dec_strategies.py#L82) I guess we should eventually call instead `env.get_num_starts` and `env.select_start_nodes` instead of having to go through the default function from `ops` (i guess this is what you meant in the  discussion!)
- [Minor] we should also add some docstring
CC: https://github.com/ai4co/rl4co/discussions/117

File: rl4co/envs/scheduling/ffsp.py [hunk 2]
```diff
@@ -333,11 +421,18 @@ def generate_data(self, batch_size) -> TensorDict:
...
+            low=self.min_time,
+            high=self.max_time,
+            size=(*batch_size, self.num_job, self.num_machine, self.num_stage),
+        ).to(self.device)
+
+        if self.flatten_stages:
+            run_time = (
+                run_time.transpose(-2, -1)
+                .contiguous()
+                .view(*batch_size, self.num_job, self.num_machine_total)
```
Comment (fedebotu) [inline rl4co/envs/scheduling/ffsp.py:434]: [Comment only] Great, these `view` parts are always tricky... I wish there was a standardized way to do this

File: rl4co/envs/scheduling/ffsp.py [hunk 3]
```diff
@@ -346,5 +441,105 @@ def generate_data(self, batch_size) -> TensorDict:
             batch_size=batch_size,
         )
 
-    def render(self, td: TensorDict):
-        raise NotImplementedError("TODO: render is not implemented yet")
+    def render(self, td: TensorDict, idx: int):
```
Comment (fedebotu) [inline rl4co/envs/scheduling/ffsp.py:444]: [Comment only] awesome!

File: rl4co/models/nn/dec_strategies.py [hunk 4]
```diff
@@ -79,7 +77,10 @@ def pre_decoder_hook(self, td: TensorDict, env: RL4COEnvBase):
 
         # Multi-start decoding: first action is chosen by ad-hoc node selection
         if self.num_starts > 1:
-            action = self.select_start_nodes_fn(td, env, self.num_starts)
+            if self.select_start_nodes_fn is not None:
+                action = self.select_start_nodes_fn(td, env, self.num_starts)
+            else:
+                action = env.select_start_nodes(td, num_starts=self.num_starts)
```
Comment (fedebotu) [inline rl4co/models/nn/dec_strategies.py:83]: [Comment only] ok, I am dumb, I commented above on something you did here already :laughing:

File: rl4co/models/zoo/matnet/encoder.py [hunk 5]
```diff
@@ -51,7 +54,7 @@ def __init__(
         self.mix_W2 = nn.Parameter(mix_W2)
         self.mix_b2 = nn.Parameter(mix_b2)
 
-        self.out_proj = nn.Linear(embedding_dim, embedding_dim, bias=bias)
+        self.out_proj = nn.Linear(embedding_dim, embedding_dim, bias=True)
```
Comment (fedebotu) [inline rl4co/models/zoo/matnet/encoder.py:57]: Is there a reason why the bias is set to True? Because of the original code?

File: rl4co/models/zoo/matnet/policy.py [hunk 6]
```diff
@@ -26,36 +41,171 @@ class MatNetPolicy(AutoregressivePolicy):
...
         **kwargs,
     ):
-        if env_name not in ["atsp"]:
-            log.error(f"env_name {env_name} is not originally implemented in MatNet")
+        if env.name not in ["atsp", "ffsp"]:
+            log.error(f"env_name {env.name} is not originally implemented in MatNet")
+
+        if env.name == "ffsp":
+            decoder = MatNetFFSPDecoder(
+                env=env,
```
Comment (fedebotu) [inline rl4co/models/zoo/matnet/policy.py:59]: Is passing the full `env` necessary here? It seems that in the `FFSPDecoder` no env is needed

File: rl4co/models/zoo/matnet/policy.py [hunk 7]
```diff
@@ -26,36 +41,171 @@ class MatNetPolicy(AutoregressivePolicy):
 
     def __init__(
         self,
-        env_name: str,
+        env: RL4COEnvBase,
```
Comment (fedebotu) [inline rl4co/models/zoo/matnet/policy.py:44]: [Minor] usally we allow to pass both `str` and `RL4COEnvBase` as seen [here](https://github.com/ai4co/rl4co/blob/66622a2e7d616e72696904e1552cf7d61dad817f/rl4co/models/zoo/common/autoregressive/decoder.py#L75C1-L76C1)

**Review by fedebotu — COMMENTED** [review]
Great job :rocket:  left some comments here and there

File: rl4co/models/zoo/matnet/encoder.py (hunk 5 above)
Comment (LTluttmann) [reply rl4co/models/zoo/matnet/encoder.py:57]: oh yes, after struggling to obtain the results reported in the paper I tried to copy the architecture as much as possible, realizing that the bias in the projection of the heads does not have much effect. I would therefore revert this change.

File: rl4co/models/zoo/matnet/policy.py (hunk 6 above)
Comment (LTluttmann) [reply rl4co/models/zoo/matnet/policy.py:59]: you are right! I will remove this line

File: rl4co/models/zoo/matnet/encoder.py (hunk 5 above)
Comment (fedebotu) [reply rl4co/models/zoo/matnet/encoder.py:57]: As you wish, this is minor!

**Review by fedebotu — APPROVED** [review]
